content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+sequence of points, there are other technics that would generate almost
+perfect histograms only using a fraction of the number of points which is used
+here. See [this article]({{< relref "../../scipy/qmc-basics" >}})
+for more information on this.
+
+But why does it matter? How bad this is will depend on your application.
+If you don't, statistical properties are lost. This is hard to get because for
+most people, random is random. So taking an arbitrary draw of sample should not
+matter. And it does not **if** you replicate your analysis using different
+**random** seeds. This is the part everyone is forgetting. Whether you are
+optimizing something or calculating some statistics, you could get (un)lucky
+if you do this just once.
+
+If your application needs random numbers, then do use random numbers and play
+with the fact that they are random. You do science or anything critical
+using these randomly generated numbers, then please stay away from seeds.
+
+Some will argue that seeding is just used by developers for testing, well it is
+not. People don't like uncertainties. They naturally try to "fix" this
+"issue" by using a seed in their systems, and not just for testing.
+The worst part here is that people tend to use the same numbers. Some
+copy/pasting might be at play there as well.
+
+Think about it for a while. If everybody is using a seed, which is the same,
+and the original intent was to interact with random numbers for statistical
+analysis or optimization. Do you see the bias that is being introduced?
+
+## Testing with random numbers
+
+Before showing any code though, ask yourself if you actually need ot use
+random numbers to write your test. You might just need a small matrix with
+hard-coded numbers. Not only you don't need to run the code to know the values,
+but you will be 100% certain that this is stable on any platform/architecture
+and future-proof. You have more advanced needs, ok then maybe using known
+sequences (quasi-random numbers) is a good alternative (again see
+[here]({{< relref "../../scipy/qmc-basics" >}})
+).
+
+If you want or need to use NumPy for that, then either use
+`np.random.Generator` or the legacy `np.random.RandomState`. Aside from
+backward compatibility concerns, `np.random.RandomState` is still present in
+NumPy for testing.
+
+```python
+rng = np.random.default_rng()
+# or
+rng = np.random.RandomState()
+```
+
+[NEP19](https://numpy.org/neps/nep-0019-rng-policy.html) actually
+recommends `np.random.RandomState` for testing as the byte stream is not
+guaranteed as strongly as before with the new generators. In practice, changes
+are not happening often though. What matters in both cases is that you use the
+object based API and not the legacy approach with global state.
+
+Now, one thing with tests is that we want them to be reproducible so that
+we can inspect our code and find issues. Here you want to use a seed.
+This is the **only** valid use for a seed. (Sure if you are an expert in
```

Comment:
Reproducibility is.  But reproducibility is not confined to testing.  My gut feeling would not be to paint seeding as dark here.  But rather say clearly that adding the seeding is best done only when reproducibility becomes vital.  I.e. not, if the analysis fails half the time otherwise...